reference/titan: add titan docs (#2494)
pingcap/docs-cn: reference/titan/configuration.md

---
title: Titan 配置
category: reference
---

# Titan 配置

Titan 是基于 RocksDB 开发的存储引擎插件，通过把 key 和 value 分离存储，在 value 较大的场景下，减少写放大，降低 RocksDB 后台 compaction 对 I/O 带宽和 CPU 的占用，以提高性能。

## 开启 Titan

> Titan 对 RocksDB 兼容，也就是说，使用 RocksDB 存储引擎的现有 TiKV 实例可以直接开启 Titan。开启的方法是修改 TiKV 配置并重启 TiKV：

Reviewer: ```suggestion
Titan 对 RocksDB 兼容，也就是说，使用 RocksDB 存储引擎的现有 TiKV 实例可以直接开启 Titan。如果使用 TiUP 部署的集群，开启的方法是执行 `tiup cluster edit-config ${cluster-name}` ，编辑配置文件，开启 titan

{{< copyable "shell-regular" >}}
```shell
server_configs:
  tikv:
    rocksdb.titan.enabled: true
```
Reload 配置，同时也会在线滚动重启 tikv
`tiup cluster reload likun-single -R tikv`

也可以直接编辑 tikv 配置文件开启 titan（线上环境不推荐）：${deploy-dir}/conf/tikv.toml

> ---

Reviewer 2: 不建议用较深的 URL，建议直接采用：reference/titan-user-guide.md

Author: 我看之前涉及多个文档的部分也都是新建了个文件夹，如果直接放在 reference 下会不会有点乱

title: Titan 配置
category: reference
---

> # Titan 配置

Reviewer 3: 如果本文档为`Titan 配置`文档的话，是否有更具概括性的 Titan overview 文档呢？

Author: design 文档就算是 overview

> Titan 是基于 RocksDB 开发的存储引擎插件，通过把 key 和 value 分离存储，以在 value 较大的场景下，降低 RocksDB 后台 compaction 对 I/O 带宽和 CPU 的占用，同时提高 SSD 寿命。

Reviewer: 性能可以提高

> ## 适用场景

Reviewer: 这篇文章标题是 TiTan 配置，这里的试用场景，是不是应该放到 overview 里，这篇主要介绍各种操作和配置

> Titan 适合在以下场景中使用：

Reviewer: 在满足这些场景和条件的情况下，带来的性能提升有多少？需要一些性能数据和截图

- 前台写入量较大，RocksDB 大量触发 compaction 消耗大量 I/O 带宽或者 CPU 资源，造成 TiKV 前台读写性能较差。
- 前台写入量较大，由于 I/O 带宽瓶颈或 CPU 瓶颈的限制，RocksDB compaction 进度落后较多频繁造成 write stall。
- 前台写入量较大，RocksDB 大量触发 compaction 造成 I/O 写入量较大，影响 SSD 盘的寿命。

开启 Titan 需要考虑以下前提条件：

> - Value 较大。即 value 平均大小比较大，或者数据中大 value 的数据总大小占比比较大。目前 Titan 默认 1kb 以上大小的 value 是大 value，根据实际情况 512b 以上大小的 value 也可以看作是大 value。

Reviewer: 建议提一下 value 的上限，目前是 6m，不然会误解可以突破 tikv 6m 的限制

- 没有范围查询或者对范围查询性能不敏感。Titan 存储数据的顺序性较差，所以相比 RocksDB 范围查询的性能较差，尤其是大范围查询。在测试中 Titan 范围查询性能相比 RocksDB 下降 40% 到数倍不等。
- 磁盘剩余空间足够。Titan 降低写放大是通过牺牲空间放大达到的。另外由于 Titan 逐个压缩 value，压缩率比 RocksDB（逐个压缩 block）要差。这两个因素一起造成 Titan 占用磁盘空间比 RocksDB 要多，这是正常现象。根据实际情况和不同的配置，Titan 磁盘空间占用可能会比 RocksDB 多一倍。

## 开启 Titan

> Titan 对 RocksDB 兼容，也就是说，使用 RocksDB 存储引擎的现有 TiKV 实例可以直接开启 Titan。开启的方法是修改 TiKV 配置并重启 TiKV：

Reviewer: 后面都用 tiup 了，需要改成 tiup 中改配置的方式

Titan 对 RocksDB 兼容，也就是说，现有使用 RocksDB 存储引擎的 TiKV 实例可以直接开启 Titan。开启的方法是修改 TiKV 配置并重启 TiKV：

> 

Reviewer 3: ```suggestion

{{< copyable "" >}}

```

``` toml
[rocksdb.titan]
enabled = true
```

> 开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。可以通过观察 TiKV Details - Titan kv - blob file size 确认数据保存在 Titan 中部分的大小。

Reviewer 3: ```suggestion
开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。
```

> 开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。

Reviewer: 开启后，titan 和 rocksdb 是 2 个实例么？长期静态的数据，是否会一直留在 rocksdb 中？

> 如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction。请参看 tikv-ctl 文档。

Author: 感觉没必要，看其他地方也没列出来 tikv-ctl 文档链接

> 如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction。请参看 tikv-ctl 文档。

Reviewer: 建议加一句描述，线上环境不建议全量 compaction，对性能有影响

> 如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction。请参看 tikv-ctl 文档。

Reviewer: 那还是加下链接吧，因为这个地方很容易有加速的需求

**注意：**
>

> > RocksDB 无法读取 Titan 的数据，但用 RocksDB 打开 Titan 数据也不会造成数据损坏。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan （误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 错误。请参看[关闭 Titan](#关闭-titan实验性) 一节。

Reviewer: ```suggestion
> 在不开启 Titan 功能的情况下，RocksDB 无法读取已经迁移到 Titan 的数据。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan （误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 错误。如需要关闭 Titan，请参看[关闭 Titan](#关闭-titan实验性) 一节。
```

> ## 参数调整

Reviewer: ```suggestion
## 相关参数介绍
```

## 参数调整

```toml
[rocksdb.titan]
max-background-gc = 1
```

> Titan GC 线程数。当从 TiKV Details - Thread CPU - RocksDB CPU 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑增加 Titan GC 线程池大小。

Reviewer 3: 监控面版应该用**加粗**格式。
```suggestion
Titan GC 线程数。当从 **TiKV Details** - **Thread CPU** - **RocksDB CPU** 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑增加 Titan GC 线程池大小。
```

```toml
[rocksdb.defaultcf.titan]
min-blob-size = "1KB"
```

> 大 value 大小的阈值。当写入的 value 大小小于这个值时，value 会保存在 RocksDB 中，反之则保存在 Titan 的 blob file 中。视乎 value 大小的分布，增大这个值可以使更多 value 保存在 RocksDB，读取这些小 value 的性能会稍好一些；减少这个值可以使更多 value 保存在 Titan 中，进一步减少 RocksDB compaction。

Reviewer 3: ```suggestion
大 value 的大小阈值。当写入的 value 小于这个值时，value 会保存在 RocksDB 中，反之则保存在 Titan 的 blob file 中。根据 value 大小的分布，增大这个值可以使更多 value 保存在 RocksDB，读取这些小 value 的性能会稍好一些；减少这个值可以使更多 value 保存在 Titan 中，进一步减少 RocksDB compaction。
```

```toml
[rocksdb.defaultcf.titan]
blob-file-compression = "lz4"
```

Titan 中 value 所使用的压缩算法。Titan 中压缩是以 value 为单元的。

```toml
[rocksdb.defaultcf.titan]
blob-cache-size = 0
```

> Titan 中 value 的缓存大小。更大的缓存能提高 Titan 读性能，但过大的缓存会造成 OOM。建议在数据库稳定运行后，根据监控把 RocksDB block cache （storage.block-cache.capacity） 设置为 store size 减去 blob file size 的大小，blob-cache-size 设置为 内存大小 * 50% 减去 block cache 的大小。这是为了保证 block cache 足够缓存整个 RocksDB 的前提下，blob cache 尽量大。

Reviewer 3: 是否可以减少一下不必要的英文单词而用中文书写呢？
```suggestion
Titan 中 value 的缓存大小。更大的缓存能提高 Titan 读性能，但过大的缓存会造成 OOM。建议在数据库稳定运行后，根据监控把 RocksDB block cache (storage.block-cache.capacity) 设置为 store size 减去 blob file size 的大小，`blob-cache-size` 设置为`内存大小 * 50% 再减去 block cache 的大小`。这是为了保证 block cache 足够缓存整个 RocksDB 的前提下，blob cache 尽量大。
```

Reviewer 4: 这里英文单词很多都是Titan/RocksDB里面的专有技术名词，或者是特指监控里面的某个同名的值，用英文单词比较容易保持一致性，不容易混淆。

写放大上界 = 1 / discardable_ratio
空间放大上界 = 1 / ( 1 - discarable_ratio )
可以看到，减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。

```toml
[rocksdb]
rate-bytes-per-sec = 0
```

该选项并不是 Titan 独有的设置。该选项限制 RocksDB compaction 的 I/O 速率，以达到在流量高峰时，限制 RocksDB compaction 减少其 I/O 带宽和 CPU 消耗对前台读写性能的影响。当开启 Titan 时，该选项限制 RocksDB compaction 和 Titan GC 的 I/O 速率总和。当发现在流量高峰时 RocksDB compaction 和 Titan GC 的 I/O 和/或 CPU 消耗过大，可以根据磁盘 I/O 带宽和实际写入流量适当配置这个选项。

## 关闭 Titan（实验性）

通过设置 rocksdb.defaultcf.titan.blob-run-mode 可以关闭 Titan。blob-run-mode 可以设置为以下几个值之一：

- 当设置为 “kNormal” 时，Titan 处于正常读写的状态。
- 当设置为 “kReadnly” 时，新写入的 value 不论大小均会写入 RocksDB。
- 当设置为 “kFallback” 时，新写入的 value 不论大小均会写入 RocksDB，并且当 RocksDB 进行 compaction 时，会自动把所碰到的存储在 Titan blob file 中的 value 移回 RocksDB。

当需要关闭 Titan 时，可以设置 blob-run-mode = “kFallback”，并通过 tikv-ctl 执行全量 compaction。此后通过监控确认 blob file size 降到 0 以后，可以更改 rocksdb.titan.enabled = false 并重启 TiKV。

关闭 Titan 是实验性功能，非必要不建议使用。

## Level Merge：提升范围查询性能（实验性）

TiKV 4.0 中 Titan 提供新的算法提升范围查询性能并降低 Titan GC 对前台写入性能的影响。这个新的算法称为 level merge。Level merge 可以通过以下选项开启：

```toml
[rocksdb.defaultcf.titan]
level-merge = true
```

> 开启 level merge 的好处包括：

Reviewer: 这几个好处都缺少性能指标，或者说量化，无法让用户决定是否开启，建议加上

> - 大幅提升 Titan 的范围查询性能。

Reviewer: 与 原生 rocksdb 相比如何？

    {{< copyable "shell-regular" >}}

    ```shell
    `tiup cluster reload ${cluster-name} -R tikv`
    ```

    具体命令，可参考[通过 TiUP 修改配置参数](/how-to/maintain/tiup-operations.md#修改配置参数)。

+ 方法二：直接编辑 TiKV 配置文件开启 Titan（线上环境不推荐)。

    {{< copyable "" >}}

    ``` toml
    [rocksdb.titan]
    enabled = true
    ```

开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。

如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction，具体参考[手动 compact](/reference/tools/tikv-control.md###手动-compact-整个-TiKV-集群的数据)。

> **注意：**
>
> 在不开启 Titan 功能的情况下，RocksDB 无法读取已经迁移到 Titan 的数据。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan（误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 错误。如需要关闭 Titan，参考[关闭 Titan](#关闭-titan实验性) 一节。

## 相关参数介绍

使用 TiUP 调整参数，请参考[修改配置参数](/how-to/maintain/tiup-operations.md#修改配置参数)。

+ Titan GC 线程数。

    当从 **TiKV Details** - **Thread CPU** - **RocksDB CPU** 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑增加 Titan GC 线程池大小。

    {{< copyable "" >}}

    ```toml
    [rocksdb.titan]
    max-background-gc = 1
    ```

> + 大 value 的大小阈值。

Reviewer 3: ```suggestion
+ value 的大小阈值。
```
